### PR TITLE
Add project telemetry

### PR DESCRIPTION
**Pull request by andy-ms**
This adds a telemetry event the first time a project is opened in a session.
The event contains a count of files by extension and scrubbed compiler options.

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -1704,4 +1707,47 @@ namespace ts {
     function caseInsensitiveKeyMapper(key: string) {
         return key.toLowerCase();
     }
+
+    /**
+     * Produces a cleaned version of compiler options with personally identifiying info (aka, paths) removed.
+     * Also converts enum values back to strings.
+     */
+    export function convertCompilerOptionsForTelemetry(opts: ts.CompilerOptions): ts.CompilerOptions {
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: `@internal` ?

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -29,7 +30,49 @@ namespace ts.server {
...
+        readonly data: ProjectTelemetryEventData;
+    }
+
+    export interface FileStats {
+        readonly ts: number;
+        readonly js: number;
+        readonly dts: number;
+    }
+
+    function getFileStats(fileNames: string[]): FileStats {
```
Comment (mhegazy) [inline src/server/editorServices.ts]: why not just use `countEachFileTypes`

**Comment by mhegazy** [comment]
We also should add the new event in protocol.ts

**Comment by andy-ms** [comment]
protocol.ts already has a "telemetry" event, should we use that? Or change the event name to "projectTelemetry"?

**Comment by mhegazy** [comment]
> protocol.ts already has a "telemetry" event, should we use that? Or change the event name to "projectTelemetry"?

i would add a new one. @mjbvz any preferences here?

**Comment by mhegazy** [comment]
@andy-ms we should also set `version` on the telemetry event that would be `ts.version`. this way we can tell what version of the server sent it. we do something similar with `typingsInstallerVersion`.

File: src/server/editorServices.ts [hunk 3]
```diff
@@ -1029,9 +1037,19 @@ namespace ts.server {
             const data: ProjectInfoTelemetryEventData = {
```
Comment (mhegazy) [inline src/server/editorServices.ts]: for completeness lets also include `include`, `exclude` and `compileOnSave`

**Comment by mjbvz** [comment]
@mhegazy It may be more consistent to stick with a single `telemetry` event and then use `telemetryEventName` to differentiate the payload type. We already do this for the `typingsInstalled` telemetry event

File: src/server/editorServices.ts [hunk 4]
```diff
@@ -29,7 +30,47 @@ namespace ts.server {
...
+        readonly compileOnSave: boolean;
+        readonly typeAcquisition: ProjectInfoTypeAcquisitionData;
+        /** TypeScript version used by the server. */
+        readonly version: string;
+    }
+
+    export interface ProjectInfoTypeAcquisitionData {
+        readonly enable: boolean;
+        // Actual values of include/exclude entries are scrubbed.
+        readonly include: Array<"">;
```
Comment (mhegazy) [inline src/server/editorServices.ts]: should not this be like `include` and `exclude` and be  `boolean | undefined`?

Comment (andy-ms) [inline src/server/editorServices.ts]: In contrast to a tsconfig "include", typeAcquisition include can apparently be provided by an external config. It also appears to be getting automatically set to an empty array if not explicitly provided.
We could make it be a `boolean` that is equal to whether the array is non-empty though.

**Comment by mhegazy** [comment]
A few more pieces of info that just came to mind: 
1. `languageServiceDisabled`
2. config file name `tsconfig.json` vs `jsconfig.json` vs other,
3. `extends` being defined in the tsconfig.json
4. is external project or configured project

File: src/server/editorServices.ts [hunk 5]
```diff
@@ -44,13 +44,18 @@ namespace ts.server {
...
         // Otherwise, we will use "true" if the property is present and "false" if it is missing.
+        readonly extends: boolean | undefined;
         readonly files: boolean | undefined;
         readonly include: boolean | undefined;
         readonly exclude: boolean | undefined;
         readonly compileOnSave: boolean;
         readonly typeAcquisition: ProjectInfoTypeAcquisitionData;
+
+        /** Is 'undefined' if config file name is not tsconfig or jsconfig. */
+        readonly configFileName: "tsconfig.json" | "jsconfig.json" | undefined;
```
Comment (mhegazy) [inline src/server/editorServices.ts]: undefined is not going to survive in the json serialization. consider adding `"other"` value.

**Pull request merged**